hubtools/src/lib.rs

```diff
@@ -1,22 +1,210 @@
 // This Source Code Form is subject to the terms of the Mozilla Public
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
+use object::{Object, ObjectSection};
 use path_slash::PathBufExt;
 use thiserror::Error;
 use zerocopy::{AsBytes, FromBytes};
 
 use std::{
-    collections::BTreeMap,
+    collections::{btree_map::Entry, BTreeMap},
     io::{Cursor, Read, Write},
     path::{Path, PathBuf},
-    process::Command,
 };
 
-/// A chunk of memory
-#[derive(Debug, Hash)]
-pub struct LoadSegment {
-    pub source_file: PathBuf,
+#[derive(Debug)]
+pub struct RawHubrisImage {
+    pub start_addr: u32,
     pub data: Vec<u8>,
+    pub kentry: u32,
+}
+
+impl RawHubrisImage {
+    pub fn from_segments(
+        data: &BTreeMap<u32, Vec<u8>>,
+        kentry: u32,
+        gap_fill: u8,
+    ) -> Result<Self, Error> {
+        // TODO: check for memory range overlaps here
+        let mut prev: Option<u32> = None;
+        let mut out = vec![];
+        for (addr, data) in data.iter() {
+            if let Some(mut prev) = prev {
+                while prev != *addr {
+                    out.push(gap_fill);
+                    prev += 1;
+                }
+            }
+            prev = Some(*addr + data.len() as u32);
+            out.extend(data);
+        }
+        let start_addr = data.keys().next().cloned().unwrap_or(0);
```

Comment:
Nit - I think

```suggestion
        let start_addr = data.keys().next().copied().unwrap_or(0);
```

is slightly clearer (when the type is copyable).

Reply:
Yup, done